libpod/container.go

```diff
@@ -808,3 +814,42 @@ func (c *Container) isStopped() (bool, error) {
 	}
 	return c.state.State == ContainerStateStopped, nil
 }
+
+// nil container state to the database
+func (c *Container) save() error {
+	if err := c.runtime.state.SaveContainer(c); err != nil {
+		return errors.Wrapf(err, "error saving container %s state", c.ID())
+	}
+	return nil
+}
+
+// CleanupStorage unmounts all mount points in container and cleans up container storage
+func (c *Container) CleanupStorage() error {
+	for _, mount := range c.mounts {
+		if err := unix.Unmount(mount, unix.MNT_DETACH); err != nil {
+			logrus.Warnf("container %s failed to unmount %s : %v", c.ID(), mount, err)
+		}
+	}
+	c.mounts = []string{}
+
+	if c.state.Mounted {
+		// TODO Remove this once --detached containers are handled
```

Comment:
This needs to be permanent if it's being called in teardownStorage() - we should always be unmounting all active mounts when we tear down the container to prepare for removal, including the SHM

Reply:
If state is not mounted then shm can not be mounted.  The idea of the TODO is that this should happen when a detached container exits, We should not need to do this.  The code above going through the c.mounts should have cleared this up.